plugins/BEdita/API/src/Middleware/AnalyticsMiddleware.php

```diff
@@ -0,0 +1,170 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2017 ChannelWeb Srl, Chialab Srl
+ *
+ * This file is part of BEdita: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * See LICENSE.LGPL or <http://gnu.org/licenses/lgpl-3.0.html> for more details.
+ */
+namespace BEdita\API\Middleware;
+
+use BEdita\Core\State\CurrentApplication;
+use BEdita\Core\Utility\LoggedUser;
+use Cake\Core\Configure;
+use Cake\Http\Response;
+use Cake\Log\Log;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Throwable;
+
+/**
+ * Middleware to trace analytics data
+ *
+ * @since 4.0.0
+ */
+class AnalyticsMiddleware
+{
+    /**
+     * Request start time
+     *
+     * @var float
+     */
+    protected $startTime = null;
+
+    /**
+     * Analytics data
+     *
+     * @var array
+     */
+    protected $data = [];
+
+    /**
+     * Registered callbacks fo analytics extensions
+     *
+     * @var array
+     */
+    protected static $callbacks = [];
+
+    /**
+     * Configure analytics logger if not configured yet
+     *
+     * @return void
+     */
+    public function __construct()
+    {
+        $this->startTime = microtime(true);
+        if (defined('TIME_START')) {
+            $this->startTime = TIME_START;
+        }
+
+        if (!in_array('analytics', Log::configured())) {
+            Log::config('analytics', [
+                'className' => 'File',
+                'path' => LOGS,
+                'scopes' => ['analytics'],
+                'file' => 'analytics',
+            ]);
+        }
+    }
+
+    /**
+     * Getter for $startTime
+     *
+     * @return float
+     * @codeCoverageIgnore
+     */
+    public function getStartTime()
+    {
+        return $this->startTime;
+    }
+
+    /**
+     * Getter for $data
+     *
+     * @return float
+     * @codeCoverageIgnore
+     */
+    public function getData()
+    {
+        return $this->data;
+    }
+
+    /**
+     * Register new analytics extension callback
+     *
+     * @param mixed $newCallback Callback to register
+     * @return void
+     */
+    public static function registerCallback($newCallback)
+    {
+        if (!is_callable($newCallback)) {
+            Log::warning('Bad callback ' . print_r($newCallback, true));
+
+            return;
+        }
+        static::$callbacks[] = $newCallback;
+    }
+
+    /**
+     * Get analytics custom data from registered callbacks
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @return array
+     */
+    protected function readCallbackData(ServerRequestInterface $request, ResponseInterface $response)
```

Comment:
Why not dispatching an event instead? By using `EventDispatcherInterface` and `EventDispatcherTrait` we should be able to obtain the same result using a pattern that's already common in the whole application.

Components that are willing to add their own callback could just register themselves as listeners for the event being dispatched and modifying event result.

Reply:
Right, I'll give a try